### PR TITLE
Configurable aggregate translator

### PR DESCRIPTION
**Pull request by codeliner**
See #68 

- Supports configurable method names for extracting
  identifier, pending stream events and reconstitution of the AR
- Forces AR methods to be public available to keep API of
  translator simple
- Offers a way to translate custom events from/to Messages
- AggregateTranslator interface slightly changed to prepare
  support for iterable Stream feature coming in v6.0

I've decided to not support different method visibility. The translator requires all methods to be public available. If one need more advanced translation logic he or she is asked to use a custom AggregateTranslator.

Another feature I added is that the ConfigurableAggregateTranslator takes event translation callbacks as optional arguments to support translation of custom domain events from/to `Prooph\Common\Messaging\Message`s.

Now the domain model can be fully decoupled from the event store and prooph/event-sourcing plays  only a minor role (should only be used in prooph projects and as an example implementation for other domain models)

Documentation will follow in a another PR. I assigned all documentation issues to me and will submit one big docs PR because a lot of things changed and I want to split README to keep it small and focused and move documentation into a `docs` folder.

### CODE REVIEW

**Comment by codeliner** [comment]
ping @prolic

**Comment by codeliner** [comment]
yes, coverage now at 85.507 % :tada:

File: src/Aggregate/ConfigurableAggregateTranslator.php [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+     */
+    public function __construct(
+        $identifierMethodName = null,
+        $popRecordedEventsMethodName = null,
+        $staticReconstituteFromHistoryMethodName = null,
+        $eventToMessageCallback = null,
+        $messageToEventCallback = null)
+    {
+        if (null !== $identifierMethodName) {
+            Assertion::string($identifierMethodName, 'Identifier method name needs to be a string');
```
Comment (prolic) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: Better use Assertion::minLength($identifierMethodName, 1, 'message');

File: src/Aggregate/ConfigurableAggregateTranslator.php [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            $this->identifierMethodName = $identifierMethodName;
+        }
+
+        if (null !== $popRecordedEventsMethodName) {
+            Assertion::string($popRecordedEventsMethodName, 'Pop recorded events method name needs to be a string');
```
Comment (prolic) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: same here, Assertion::minLength

File: src/Aggregate/ConfigurableAggregateTranslator.php [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+            $this->popRecordedEventsMethodName = $popRecordedEventsMethodName;
+        }
+
+        if (null !== $staticReconstituteFromHistoryMethodName) {
+            Assertion::string($staticReconstituteFromHistoryMethodName, 'Method name for static method reconstitute from history needs to be string');
```
Comment (prolic) [inline src/Aggregate/ConfigurableAggregateTranslator.php]: same here, Assertion::minLength

File: tests/Aggregate/ConfigurableAggregateTranslatorTest.php [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        $this->assertEquals('123', $translator->extractAggregateId($ar->reveal()));
+    }
+
+    /**
+     * @test
+     * @expectedException \InvalidArgumentException
+     */
+    public function it_forces_identifier_method_name_to_be_a_string()
+    {
+        new ConfigurableAggregateTranslator(666);
```
Comment (prolic) [inline tests/Aggregate/ConfigurableAggregateTranslatorTest.php]: 666? Are you satanic? :)

File: tests/Aggregate/ConfigurableAggregateTranslatorTest.php [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+use Prooph\EventStoreTest\Mock\DefaultAggregateRoot;
+use Prooph\EventStoreTest\Mock\DefaultAggregateRootContract;
+use Prooph\EventStoreTest\Mock\FaultyAggregateRoot;
+use Prooph\EventStoreTest\TestCase;
+
+/**
+ * Class ConfigurableAggregateTranslatorTest
+ *
+ * @package Prooph\EventStoreTest\Aggregate
+ * @author Alexander Miertsch <alexander.miertsch.extern@sixt.com>
```
Comment (prolic) [inline tests/Aggregate/ConfigurableAggregateTranslatorTest.php]: Now a sixt email address?

File: tests/Aggregate/ConfigurableAggregateTranslatorTest.php [hunk 6]
```diff
@@ -0,0 +1,287 @@
+<?php
+/*
+ * This file is part of the prooph/event-store.
+ * (c) 2014-2015 prooph software GmbH <contact@prooph.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * Date: 8/22/15 - 9:42 PM
```
Comment (prolic) [inline tests/Aggregate/ConfigurableAggregateTranslatorTest.php]: Date format in copyright should be consistent, you use always different formats ;)

File: tests/Mock/DefaultAggregateRoot.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+{
+    private $historyEvents = [];
+
+    /**
+     * @param Message[] $historyEvents
+     * @return DefaultAggregateRootContract
+     */
+    public static function reconstituteFromHistory($historyEvents)
+    {
+        $self = new self();
```
Comment (prolic) [inline tests/Mock/DefaultAggregateRoot.php]: Perhaps a private constructor is nicer, but that's only a test mock, so I am okay with this.

**Comment by prolic** [comment]
Only small remarks, can be merged once fixed.

File: tests/Aggregate/ConfigurableAggregateTranslatorTest.php (hunk 4 above)
Comment (codeliner) [inline tests/Aggregate/ConfigurableAggregateTranslatorTest.php]: yeah, an evil number to evoke an exception ;)

File: tests/Aggregate/ConfigurableAggregateTranslatorTest.php (hunk 5 above)
Comment (codeliner) [inline tests/Aggregate/ConfigurableAggregateTranslatorTest.php]: damn, sixt is everywhere ;)

File: tests/Aggregate/ConfigurableAggregateTranslatorTest.php (hunk 6 above)
Comment (codeliner) [inline tests/Aggregate/ConfigurableAggregateTranslatorTest.php]: seems to be system dependent. Changed all dates to mm/dd/yy

File: tests/Mock/DefaultAggregateRoot.php (hunk 7 above)
Comment (codeliner) [inline tests/Mock/DefaultAggregateRoot.php]: in a real model absolutely

**Pull request merged**